spring-integration-mqtt/src/main/java/org/springframework/integration/mqtt/inbound/AbstractMqttMessageDrivenChannelAdapter.java

```diff
@@ -70,25 +79,54 @@ public abstract class AbstractMqttMessageDrivenChannelAdapter extends MessagePro
 
 	private MqttMessageConverter converter;
 
-	protected final Lock topicLock = new ReentrantLock(); // NOSONAR
+	private ClientManager.ConnectCallback clientManagerConnectCallback;
 
 	public AbstractMqttMessageDrivenChannelAdapter(@Nullable String url, String clientId, String... topic) {
 		Assert.hasText(clientId, "'clientId' cannot be null or empty");
-		Assert.notNull(topic, "'topics' cannot be null");
-		Assert.noNullElements(topic, "'topics' cannot have null elements");
 		this.url = url;
 		this.clientId = clientId;
-		this.topics = new LinkedHashSet<>();
+		this.topics = initTopics(topic);
+		this.clientManager = null;
+	}
+
+	public AbstractMqttMessageDrivenChannelAdapter(ClientManager<T, C> clientManager, String... topic) {
+		Assert.notNull(clientManager, "'clientManager' cannot be null");
+		this.clientManager = clientManager;
+		this.topics = initTopics(topic);
+		this.url = null;
+		this.clientId = null;
+	}
+
+	private Set<Topic> initTopics(String[] topic) {
+		Assert.notNull(topic, "'topics' cannot be null");
+		Assert.noNullElements(topic, "'topics' cannot have null elements");
+		final Set<Topic> initialTopics = new LinkedHashSet<>();
+		int defaultQos = 1;
 		for (String t : topic) {
-			this.topics.add(new Topic(t, 1));
+			initialTopics.add(new Topic(t, defaultQos));
 		}
+		return initialTopics;
 	}
 
 	public void setConverter(MqttMessageConverter converter) {
 		Assert.notNull(converter, "'converter' cannot be null");
 		this.converter = converter;
 	}
 
+	@Nullable
+	protected ClientManager<T, C> getClientManager() {
+		return this.clientManager;
+	}
+
+	@Nullable
+	protected ClientManager.ConnectCallback getClientManagerCallback() {
+		return this.clientManagerConnectCallback;
+	}
+
+	protected void setClientManagerCallback(ClientManager.ConnectCallback clientManagerConnectCallback) {
```

Comment:
That's not. 
The `AbstractMqttMessageDrivenChannelAdapter` creates such a `ConnectCallback`  internally and register it with the provided `ClientManager`.
See, for example, an `AbstractRSocketConnector` and its `addEndpoint()` call from the `RSocketInboundGateway`.
No reason to have some extra API around: we just call `getClientManager().addCallback()` from target channel adapter impl.
I don't see, though, how you connect two of them here...

Reply:
This setter and field is just to be able to save the listener instance to remove it later: see in `destroy` method,
```java
if (clientManager != null) {
	clientManager.removeCallback(getClientManagerCallback());
}
```

Reply:
I think both approaches are not ideal 🙂  - but I guess having protected getters/setters is kinda an overkill for it.
I like your approach with (this) better. I will implement it.